Add safe parsing without exceptions

NeQuissimus opened the pull request:
PR for #1590

/cc @fommil

fommil commented:
I think something even simpler would be better... just throwing away the error message and returning `Maybe[A]`. And replacing the existing syntax. For simple primitives I don't think we really need to see a detailed breakdown about why parsing failed and it can be added retrospectively.

NeQuissimus commented:
There should be a `toOption` and `toMaybe` on `Either`.
I would prefer keeping the error messages. I have used it to decide between truly bad input or just a required numeric widening (Int -> Long for example).

fommil commented:
perhaps these can more easily replace the existing syntax in that case, which is currently using `Validation[Throwable, A]`. This is an improvement over that.

NeQuissimus commented:
Oh, are we able to change the API?

fommil commented:
@NeQuissimus in 7.3, yup :smile:

fommil commented on core/src/main/scala/scalaz/std/String.scala:
```diff
@@ -38,33 +38,32 @@ trait StringFunctions {
+    try {
+      Success(f(s))
+    } catch {
+      case _: NumberFormatException if intChars(s) => Failure(s"${s} is outside of range for ${t} (${r._1} - ${r._2})")
```
won't this be true if `s` is empty?

fommil commented on core/src/main/scala/scalaz/std/String.scala:
```diff
@@ -38,33 +38,32 @@ trait StringFunctions {
 
-  def parseBoolean(s:String): Validation[IllegalArgumentException, Boolean] =
-    Validation.fromTryCatchThrowable[Boolean, IllegalArgumentException](s.toBoolean)
+  private def asNumber[T](f: String => T, b: String => Boolean, r: Tuple2[T, T], s: String, t: String): Validation[String, T] =
```
you can use `ClassTag` instead of `t: String` and get the primitive name expanded for free.

NeQuissimus replied:
Ah, yes, I was trying `TypeTag` and that wasn't making me happy...

fommil replied:
Generally there is need for a new `Describable` typeclass with a macro that generates the refined type name. But `ClassTag` is the closest thing we have from stdlib.

fommil commented on core/src/main/scala/scalaz/syntax/std/StringOps.scala:
```diff
@@ -24,23 +24,23 @@ final class StringOps(private val self: String) extends AnyVal {
+  def parseBigInt: Validation[String, BigInt] = s.parseBigInt(self)
 
-  def parseBigDecimal: Validation[NumberFormatException, BigDecimal] = s.parseBigDecimal(self)
+  def parseBigDecimal: Validation[String, BigDecimal] = s.parseBigDecimal(self)
```
I'm still in favour or removing the `scala.BigDecimal` constructor and replacing with the java variant (noting the scala bug in comments).

fommil commented:
you forgot to delete the syntax for the "bad" parsers

```
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/syntax/std/StringOps.scala:41:51: value parseBigInt is not a member of object scalaz.std.string
[error]   def parseBigInt: Validation[String, BigInt] = s.parseBigInt(self)
[error]                                                   ^
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/syntax/std/StringOps.scala:43:59: value parseBigDecimal is not a member of object scalaz.std.string
[error]   def parseBigDecimal: Validation[String, BigDecimal] = s.parseBigDecimal(self)
[error]                                                           ^
```

Adding some tests would be :100:

fommil commented on core/src/main/scala/scalaz/std/String.scala:
```diff
@@ -38,33 +38,32 @@ trait StringFunctions {
+    } catch {
+      case _: NumberFormatException if intChars(s) => Failure(s"${s} is outside of range for ${t} (${r._1} - ${r._2})")
+      case _: NumberFormatException => Failure(s"${s} does not represent a valid ${t}")
+      case e: Throwable => Failure(e.getMessage)
```
apparently `NonFatal` is better... (personally I think if you're making decisions based on whether you catch or don't catch OOM, you're fucked anyway)

NeQuissimus commented:
- Switched to the `java.math` variants
- Added tests, which required me to add a `Show[BigDecimal]` and `Equal[BigDecimal]` for `java.math.BigDecimal` as well as simplistic `Arbitrary[BigDecimal]` and `Arbitrary[BigInteger]`
